The Creampie Throw Update

Mercenaryblue opened the pull request:
![creampiemasks2](https://cloud.githubusercontent.com/assets/6573763/15029162/6ce9881c-121a-11e6-836c-a20aae26c98a.png)

When you throw a banana cream pie at your target, it now knock them down, then add a delicious creamy overlay.
You can clean the banana cream pie off with various method, such as soap, cleaner spray, shower.
Compatible with various mobs, such as humans, lizard-men, aliens, corgis, monkeys, etc.
Head of Security is worth double the points! Honk!

:cl:
rscadd: Throwing a Banana Cream Pie will now knock down and cream the target's face.
rscadd: You can clean off the cream with the usual methods, such as soap, shower, cleaner spray, etc.
rscadd: The Clown Federation would like to remind the crew that the HoS is worth double points! Honk!
/:cl:

Edit: Posted **Human-only version.** Will expand the project for the other mobs at a later date.

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master' into Creampie

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

Mercenaryblue committed:
Items washing banana cream

Allow common methods of cleaning to wash banana cream pie off your face.

Mercenaryblue committed:
Janiborg cream washing

Janiborgs can now wash cream by running over people.

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Creampie

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

Mercenaryblue committed:
Added Creampie Overlay code

When a banana cream pie hits a target mob, it applies a new overlay to
the mob. This overlay can then be cleaned via soap, shower, cleaner
spray, etc.

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Creampie

Mercenaryblue committed:
Cream Pie Throw Update

When hit by a banana cream pie:
Adds a creamy overlay to various mobs.
Knock targets off for less than a second.
Enable cleaning the overlay via soap, shower, and others means.

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Creampie

Zilenan91 commented:
Can you make damp rags clean people too

Niknakflak commented:
What is up with Ian's sprite in those pictures

Mercenaryblue commented:
That is not Ian

And yea, sure, damp rags. On it.

Niknakflak commented:
What is it then

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master' into Creampie

Mercenaryblue commented:
Damp Rags were already included.

Oh, that's my next project: Ein

RemieRichards commented:
Ein is a regular corgi though, not dark chocolate brown.

optimumtact commented on code/modules/food&drinks/food/snacks_pie.dm:
```diff
@@ -27,7 +27,41 @@
 		reagents.reaction(hit_atom, TOUCH)
-		qdel(src)
+
+		if(!ismob(hit_atom)) //if the creampie misses
```
why not do an is else to avoid having to check ismob twice.

optimumtact commented on code/game/atoms.dm:
```diff
@@ -312,6 +312,8 @@ var/list/blood_splatter_icons = list()
 		blood_DNA = null
 		return 1
 
+/atom/proc/wash_cream()
```
just a note but if we get any more of these it's gonna start making sense to have some kind of overlays/datum system to handle it.

optimumtact commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -733,6 +733,9 @@
 						H.update_inv_shoes()
 			M.clean_blood()
+			M.wash_cream()
+		else if(!iscarbon(M))
```
The else if here does not make sense, just use else

the if(iscarbon(M) earlier already checked for them being carbon, doing a second if for them not being carbon (which already know to be true) is inefficient and doesn't give you any extra benefit.

Mercenaryblue replied:
Yeah, that makes sense. Quick test and I'll fix it.

Mercenaryblue committed:
else fix

replace unnecessary if w/ else

Mercenaryblue committed:
Revert "else fix"

This reverts commit 5dc5707981f8bc312b8e3ac4ce9ab16e7f20c855.

Mercenaryblue committed:
else fix

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master' into Creampie

ike709 commented:
https://github.com/tgstation/-tg-station/pull/6000

Core0verload commented:
Can you make it apply reagents to target too, like eggs do?

phil235 commented on code/modules/food&drinks/food/snacks_pie.dm:
```diff
@@ -27,7 +27,42 @@
+				creamoverlay.icon_state = "creampie_drone"
+			else if(istype(H, /mob/living/simple_animal/pet/dog/corgi))
+				creamoverlay.icon_state = "creampie_corgi"
+			else return //if the target mob isn't compatible
```
that is a LOT of istypes...

phil235 commented on code/modules/food&drinks/food/snacks_pie.dm:
```diff
@@ -27,7 +27,42 @@
+			visible_message("<span class='userdanger'>[H] was creamed by the [src]!!</span>")
+			H.overlays += creamoverlay
+
+		update_icon()
```
what's the point of updating the icon when we're qdeling the item right after that?

phil235 commented on code/modules/food&drinks/food/snacks_pie.dm:
```diff
@@ -27,7 +27,42 @@
+			H.overlays += creamoverlay
+
+		update_icon()
+	qdel(src)
```
This is getting called even when the pie is caught, which should not happen.

phil235 commented:
>Can you make it apply reagents to target too, like eggs do?

It already does

Mercenaryblue commented:
Alright. Thanks for the feedback Phil, I'll test and update shortly.

Mercenaryblue committed:
Catch Fix

Catching a thrown cream pie will no longer delete the item in your hands

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

TrustyGun commented:
this should give the target some sort of creamy overlay on them

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Creampie

phil235 commented on code/modules/food&drinks/food/snacks_pie.dm:
```diff
@@ -27,7 +27,41 @@
+
+			if(istype(H, /mob/living/carbon/human))
+				var/mob/living/carbon/human/S = hit_atom
+				if(S.dna && S.dna.species && ("tail_lizard" in S.dna.species.mutant_bodyparts))
```
why check for a tail specifically? why not check the species directly instead?

Core0verload replied:
AFAIK you can attach ligger tails to humans surgically.

phil235 replied:
yeah but the creampie overlay doesn't apply on the tail.

phil235 commented on code/modules/food&drinks/food/snacks_pie.dm:
```diff
@@ -27,7 +27,41 @@
+			else if(istype(H, /mob/living/carbon/monkey))
+				creamoverlay.icon_state = "creampie_monkey"
+			else if(istype(H, /mob/living/carbon/alien))
+				creamoverlay.icon_state = "creampie_xenomorph"
```
Did you test how it looks when a xeno is resting/unconscious/dead ?

Mercenaryblue replied:
Erk. No, I haven't. Thanks for reminding me.

drynwyn commented:
Why would a cream pie knock someone over?

Mercenaryblue commented:
> Why would a cream pie knock someone over
Design choice. It draws the eye to the target's sprite, forcing it to notice the change.
Plus, it provides the clown with an ranged defensive ability. The short stun is not enough to be a threat, but it may allow you to escape with your life.

Mercenaryblue committed:
Xeno State Creamy Fix

Tossing a cream pie will now apply the appropriate overlay to a dead or
sleeping xenomorph.

Unfortunately, the xenomorph is still able to shake the cream off by
switching state, and it keep pushing the wrong overlay on critical
xenomorphs.

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master' into Creampie

Mercenaryblue commented:
Alright. I've made the script check the alien's state and apply the right overlay. Sleep and death are working well, crit overlay is messed up. And unfortunately, they still reset when the alien change state.

Still, progress was made.
I'll try to build a new datum to reduce the length of the banana cream pie file tonight.
Then I'll check if I can't make a proper overlay override for the xenomorph.

phil235 commented:
I'm not a big fan of the code, it's adding a lot of checks and proc calls just for a very small feature.
Also, the xenos would still need to update their cream overlay when changing their stance to resting/unconscious/dead/standing. Do we really want to run some code to update a creampie overlay every time a xeno stands/lies down?

Mercenaryblue commented:
I'm fine to change whatever is needed as long the functionality remains the same.

Mercenaryblue committed:
Human-only code

Simplify code to affect only humanoids.

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master' into Creampie

Mercenaryblue committed:
Revert "Human-only code"

This reverts commit f74cb0e3af08fc0b793987ed9711dbbfd4d8561b.

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

Mercenaryblue committed:
Human Code Only

Pls no Travis

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Creampie

phil235 commented on code/modules/food&drinks/food/snacks_pie.dm:
```diff
@@ -25,9 +25,28 @@
+			qdel(src)
+			return
+
+		else
```
```
	if(ishuman(hit_atom)
		var/mob/living/carbon/human/H = hit_atom
		var/image/creamoverlay etc...
		etc...
	qdel(src)
```

Mercenaryblue replied:
That qdel(src) will probably make the pie delete when caught.

phil235 commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -733,6 +733,8 @@
 						H.update_inv_shoes()
 			M.clean_blood()
+			M.wash_cream()
+		else M.wash_cream()
```
```
	H.wash_cream()
M.clean_blood()
```

since only humans gets cream now.

phil235 commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -851,6 +851,14 @@
+	var/image/creamoverlay = image('icons/effects/creampie.dmi')
+	var/mob/living/carbon/human/H = src
+	if(H.dna && H.dna.species && ("tail_lizard" in H.dna.species.mutant_bodyparts))
+		creamoverlay.icon_state = "creampie_lizard"
```
why do you absolutely want to check for the lizard's tail? DOes the overlay cover the tail? If not, you can just check H.dna.species.id == "lizard" , it's faster.

phil235 commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -851,6 +851,14 @@
+	var/mob/living/carbon/human/H = src
+	if(H.dna && H.dna.species && ("tail_lizard" in H.dna.species.mutant_bodyparts))
+		creamoverlay.icon_state = "creampie_lizard"
+	else creamoverlay.icon_state = "creampie_human"
```
```
else
	creamoverlay.icon_state = "creampie_human"
```

Mercenaryblue committed:
Optimize code

As requested

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master' into Creampie

Mercenaryblue committed:
Bug Fix

Overlay no longer unwashable when species get changed via
badmin/magic/etc.

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Creampie

Ultimate-Chimera commented:
Knocking people down? Meh... Till that's removed. :-1:
